tsdb/wlog/watcher.go

```diff
@@ -475,11 +473,46 @@ func (w *Watcher) garbageCollectSeries(segmentNum int) error {
 	return nil
 }
 
-// Read from a segment and pass the details to w.writer.
-// Also used with readCheckpoint - implements segmentReadFn.
-func (w *Watcher) readSegment(r *LiveReader, segmentNum int, tail bool) error {
+// readSegmentSeries reads the series records into w.writer from a segment.
+// It returns the EOF error if the segment is corrupted or partially written.
+func (w *Watcher) readSegmentSeries(r *LiveReader, segmentNum int) error {
 	var (
-		dec                   = record.NewDecoder(labels.NewSymbolTable()) // One table per WAL segment means it won't grow indefinitely.
+		dec    = record.NewDecoder(labels.NewSymbolTable()) // One table per WAL segment means it won't grow indefinitely.
+		series []record.RefSeries
+	)
+	for r.Next() && !isClosed(w.quit) {
+		rec := r.Record()
+		w.recordsReadMetric.WithLabelValues(dec.Type(rec).String()).Inc()
+
+		switch dec.Type(rec) {
+		case record.Series:
+			series, err := dec.Series(rec, series[:0])
+			if err != nil {
+				w.recordDecodeFailsMetric.Inc()
+				return err
+			}
+			w.writer.StoreSeries(series, segmentNum)
+		case record.Unknown:
+			// Could be corruption, or reading from a WAL from a newer Prometheus.
+			w.recordDecodeFailsMetric.Inc()
+		default:
+			// We're not interested in other types of records.
+		}
+	}
+	if err := r.Err(); err != nil {
+		return fmt.Errorf("segment %d: %w", segmentNum, err)
+	}
+	return nil
+}
+
+// readSegment reads all known records into w.writer from a segment.
+// It returns the EOF error if the segment is corrupted or partially written.
+func (w *Watcher) readSegment(r *LiveReader, startT int64, segmentNum int) (err error) {
+	var (
+		// One table per WAL segment means it won't grow indefinitely.
+		dec = record.NewDecoder(labels.NewSymbolTable())
+
+		// TODO(bwplotka): Consider zeropools.
```

Comment:
Pool should be worse, since we are already reusing memory and there is only ever zero or one slice of each type in use.